File: Graph/Minimum Obstacle Removal to Reach Corner/solution_bfs.py
class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        dirs = [(0,1), (1,0), (0,-1), (-1,0)]
        dq = deque([(0, 0, 0)])
        m, n = len(grid), len(grid[0])
        grid[0][0] = 2
        
        while dq:
            r, c, s = dq.popleft()
            if r == m-1 and c == n-1:
                return s
            for d in dirs:
                nr, nc = r + d[0], c + d[1]
                if nr < 0 or nc < 0 or nr >= m or nc >= n or grid[nr][nc] == 2:
                    continue
                if grid[nr][nc] == 1:
                    dq.append((nr, nc, s + 1))
                else:
                    dq.appendleft((nr, nc, s))
                grid[nr][nc] = 2
        return -1

File: Graph/*Minimum Cost to Make at Least One Valid Path in a Grid/solution_bfs.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& grid) {
        vector<pair<int, int>> ds{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, m*n));
        deque<pair<int, int>> dq;
        dp[0][0] = 0;
        dq.emplace_back(0, 0);
        while(!dq.empty()) {
            auto [r, c] = dq.front();
            if(r == m-1 && c == n-1) return dp[r][c];
            dq.pop_front();
            {
                auto [dr, dc] = ds[grid[r][c] - 1];
                int nr = r + dr;
                int nc = c + dc;
                if(nr < 0 || nr >= m || nc < 0 || nc >= n) {
                    ;
                } else if(dp[nr][nc] > dp[r][c]) {
                    dp[nr][nc] = dp[r][c];
                    dq.emplace_front(nr, nc);
                } 
            }
            
            for(auto &[dr, dc]: ds) {
                if(make_pair(dr, dc) == ds[grid[r][c]-1]) continue;
                int nr = r + dr;
                int nc = c + dc;
                if(nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
                if(dp[nr][nc] > dp[r][c] + 1) {
                    dp[nr][nc] = dp[r][c] + 1;
                    dq.emplace_back(nr, nc);
                }
            }
        }

        return -1;
    }
};
